libgit2: drop all old versions

SuperSandro2000 opened the pull request:
## Description of changes

Surprisingly the tools all compile with the newest version right now.

Closes #288064
Resolves the root cause that would be fixed by #289802 which is now draft for something else

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 committed:
nix-init: unpin libgit2

SuperSandro2000 committed:
biome: unpin libgit2

SuperSandro2000 committed:
cargo-generate: unpin libgit2

SuperSandro2000 committed:
cargo-workspaces: drop unused libgit2 dependency

It is using gix and libgit2{,-sys} is not in the lockfile
https://github.com/pksunkara/cargo-workspaces/blob/v0.3.1/cargo-workspaces/Cargo.lock#L680

SuperSandro2000 committed:
cargo-dephell: unpin libgit2

SuperSandro2000 committed:
libgit2_1_6: drop

SuperSandro2000 committed:
libgit2_1_3_0: drop

SuperSandro2000 committed:
cargo-unused-features: unpin libgit2

SuperSandro2000 committed:
cargo-ui: unpin libgit2

SuperSandro2000 committed:
projectable: unpin libgit2

SuperSandro2000 committed:
git-dive: unpin libgit2

SuperSandro2000 committed:
git-mit: unpin libgit2

SuperSandro2000 committed:
cargo-audit: drop unused libgit2 dependency

See https://github.com/rustsec/rustsec/pull/925

SuperSandro2000 committed:
cargo-codspeed: unpin libgit2

SuperSandro2000 committed:
cargo-update: unpin libgit2

SuperSandro2000 committed:
libgit2_1_5: drop

AndersonTorres commented:
A better commit message would be `cargo-update: unpin libgit2`.

kirillrdy reviewed (approved):
@SuperSandro2000 great stuff, always good to remove things

Result of `nixpkgs-review pr 288149` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>bat-extras.prettybat</li>
    <li>biome</li>
    <li>cargo-audit</li>
    <li>cargo-codspeed</li>
    <li>cargo-generate</li>
    <li>cargo-ui</li>
    <li>cargo-unused-features</li>
    <li>cargo-update</li>
    <li>cargo-workspaces</li>
    <li>git-dive</li>
    <li>git-mit</li>
    <li>nix-init</li>
    <li>projectable</li>
    <li>shfmt</li>
    <li>vscode-extensions.foxundermoon.shell-format</li>
  </ul>
</details>

<hr/>

Result of `nixpkgs-review pr 288149` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>cargo-generate</li>
  </ul>
</details>



<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>bat-extras.prettybat</li>
    <li>biome</li>
    <li>cargo-audit</li>
    <li>cargo-codspeed</li>
    <li>cargo-ui</li>
    <li>cargo-unused-features</li>
    <li>cargo-update</li>
    <li>cargo-workspaces</li>
    <li>git-dive</li>
    <li>git-mit</li>
    <li>nix-init</li>
    <li>projectable</li>
    <li>shfmt</li>
  </ul>
</details>

SuperSandro2000 commented:
@ofborg build cargo-generate

AndersonTorres commented:
Was `cargo-generate` working before?

@ofborg build cargo-generate

turboMaCk commented:
I believe cargo generate was working. But I would not block update on it if it now fails. From past experience with upstream they tend to be a bit behind in terms of versions compare to nixpkgs but they accept patches I've opened in past.

SuperSandro2000 commented:
Works now 😄

sternenseemann reviewed (dismissed):
Not that surprising actually. All rust programs will use `libgit2-sys` which checks for a [certain version of libgit2](https://github.com/rust-lang/git2-rs/blob/master/libgit2-sys/build.rs). If it doesn't find it, the vendored libgit2 source is built from scratch.

SuperSandro2000 commented:
Are we setting the LIBGIT2_NO_VENDOR variable in the builder? If not it is completely in transparent what we are using and this could be broken without us noticing for a long time.

sternenseemann commented:
Only here:

```
./pkgs/applications/misc/terminal-typeracer/default.nix:  LIBGIT2_NO_VENDOR = 1;
./pkgs/applications/version-management/gex/default.nix:  LIBGIT2_NO_VENDOR = 1;
```

SuperSandro2000 commented:
I've started to add the LIBGIT2_NO_VENDOR env to some packages and continue that in the next hours/days.

sternenseemann commented:
It would be nice if this variable was passed by default in `buildRustPackage` unless you explicitly opt out (and similar for the corresponding with openssl and other `*-sys` crates). We can't really expect maintainers to know about this footgun and never forget to include them…

SuperSandro2000 commented:
I tried something like that before and it ended up being a very big PR with very little interest in, so I would leave that open for someone else. When we have the env set for many packages, it should be even easier to set the default and then remove all the explizit envs from the packages.

sternenseemann commented on pkgs/top-level/aliases.nix:
```diff
@@ -1081,6 +1081,7 @@ mapAliases ({
   trezor_agent = trezor-agent; # Added 2024-01-07
   trustedGrub = throw "trustedGrub has been removed, because it is not maintained upstream anymore"; # Added 2023-05-10
   trustedGrub-for-HP = throw "trustedGrub-for-HP has been removed, because it is not maintained upstream anymore"; # Added 2023-05-10
+  turbogit = throw "turbogit has been removed, as it depends on old version of libgit2 that got removed"; # Added 2024-02-11
```
I'd split this change out and discuss it separately.

SuperSandro2000 replied:
The [4.0.0](https://github.com/b4nst/turbogit/releases/tag/v4.0.0) release was over 1.5 years ago and the [5.0.0](https://github.com/b4nst/turbogit/releases/tag/v5.0.0) was almost a year ago. The package also didn't [receive much activity since December 2021](https://github.com/NixOS/nixpkgs/issues?q=sort%3Aupdated-desc+turbogit) except treewide changes like this one and could be considered a smaller burden for libraries like libgit2.

I don't think moving this to another PR will change anything about this and just prolong that we remove old libgit2 versions. We could just remove libgit2 and mark the package broken but that would also just delay the inevitable.

I tried updating the package to a newer version but that wasn't straight forward and I gave up after 40 minutes.

sternenseemann replied:
That there is a newer release makes an even stronger case against removing it. Doesn't mean you have to update it, just that we have to wait longer for removing that one libgit2 version. It's probably affected by some CVE anyways, so stick an insecure warning into `meta` and we should be okay to move on until someone takes care of that.

SuperSandro2000 replied:
> That there is a newer release makes an even stronger case against removing it.

How that? It just proves that the package is unmaintained.

SuperSandro2000 replied:
I verified it still builds but only if you allow the insecure libgit2 package.

sternenseemann commented:
@ofborg build cargo-audit

SuperSandro2000 commented:
> ofborg build cargo-audit

FYI I've built all the packages on this branch before I rebased on master.

SuperSandro2000 commented:
Since more and more people are encountering issues with tests of the old libgit2 versions, can we merge this?

turboMaCk commented:
Personally I'm with @sternenseemann on a turbogit part of the PR (https://github.com/NixOS/nixpkgs/pull/288149#discussion_r1494872400)

Imo it would be a bit better to first role out warning before rather than hard remove. I don't expect this to be common dependency but I feel we have some responsibility toward downstream and going through even short warning / deprecation period seems for systematic to me.

This is not a hill I'm willing to die on but my vote goes towards keep legacy version for a second just so we can keep turbogit and putting warning in place and following with hard removal after (even short) period of time.

0xmangoseason commented:
> Since more and more people are encountering issues with tests of the old libgit2 versions, can we merge this?

bump

SuperSandro2000 commented:
> This is not a hill I'm willing to die on but my vote goes towards keeping legacy version for now just so we can keep turbogit and putting warning in place. We can follow with hard removal after (even short) period of time.

We cannot do this, since ofborg will complain. I also do not have increased interest in keeping libgit2 overrides just for it, since the package is basically dead and the maintainer is not taking care of it.

newAM commented:
Result of `nixpkgs-review pr 288149` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>biome</li>
    <li>cargo-audit</li>
    <li>cargo-codspeed</li>
    <li>cargo-dephell</li>
    <li>cargo-generate</li>
    <li>cargo-ui</li>
    <li>cargo-unused-features</li>
    <li>cargo-update</li>
    <li>cargo-workspaces</li>
    <li>git-dive</li>
    <li>git-mit</li>
    <li>nix-init</li>
    <li>projectable</li>
  </ul>
</details>

turboMaCk reviewed (approved):
By [popular demand](https://github.com/NixOS/nixpkgs/pull/288149#issuecomment-1952521816) I have no issues with merging this as is.

sternenseemann commented:
Rebasing a *single* commit out of this PR make this uncontroversial and immediately mergeable.

phanirithvij commented:
I apologize for butting in, but could this be merged and the turbogit be re-added later when it builds properly with an updated libgit2?

SuperSandro2000 committed:
turbogit: make libgit2 override local to package

This is the last user of an old libgit2 version which cannot
easily be updated and takes away the burden from the libgit2
maintainers.

phanirithvij commented:
@sternenseemann could you re-review and merge?

sternenseemann reviewed (changes requested):
LIBGIT2_NO_VENDOR doesn't seem to work for some reason. I noticed that cargo-audit's binary doesn't link against libgit2, so I removed it from `buildInputs` and recompiled which succeeded. So it seems that LIBGIT2_NO_VENDOR doesn't have an effect for some reason and some packages (e.g. cargo-audit) don't support the latest libgit2 version.

SuperSandro2000 commented:
That's because cargo-audit isn't using libgit2 anymore since rustsec version 0.28.0 https://github.com/rustsec/rustsec/blob/main/rustsec/CHANGELOG.md#:~:text=switched%20from%20libgit2%20to%20gitoxide%20as%20the%20git%20implementation.%20(%23925 (we are using 0.28.3, see compile log of cargo-audit) and isn't linked against it on master either

``` 
 ➜ ldd ./result/bin/cargo-audit
        linux-vdso.so.1 (0x00007ffdaebd1000)
        libssl.so.3 => /nix/store/qkfcr92mk15h8hmwzds3g5gkx0vm5l26-openssl-3.0.13/lib/libssl.so.3 (0x00007fce76ede000)
        libcrypto.so.3 => /nix/store/qkfcr92mk15h8hmwzds3g5gkx0vm5l26-openssl-3.0.13/lib/libcrypto.so.3 (0x00007fce76a5f000)
        libz.so.1 => /nix/store/bnqa606cwwff6ja8l6gz7milm7ajd1zi-zlib-1.3.1/lib/libz.so.1 (0x00007fce76a41000)
        libgcc_s.so.1 => /nix/store/9r8z1qr5jhv90fl2457qqml7qgk88fd6-gcc-13.2.0-lib/lib/libgcc_s.so.1 (0x00007fce76a1c000)
        libm.so.6 => /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/lib/libm.so.6 (0x00007fce7693a000)
        libc.so.6 => /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/lib/libc.so.6 (0x00007fce7674f000)
        /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/lib/ld-linux-x86-64.so.2 => /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/lib64/ld-linux-x86-64.so.2 (0x00007fce77f84000)
        libdl.so.2 => /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/lib/libdl.so.2 (0x00007fce7674a000)
        libpthread.so.0 => /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/lib/libpthread.so.0 (0x00007fce76745000)
```

phanirithvij commented:
@sternenseemann

phanirithvij commented:
Result of `nixpkgs-review pr 288149` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>turbogit</li>
  </ul>
</details>
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>biome</li>
    <li>cargo-audit</li>
    <li>cargo-codspeed</li>
    <li>cargo-dephell</li>
    <li>cargo-generate</li>
    <li>cargo-ui</li>
    <li>cargo-unused-features</li>
    <li>cargo-update</li>
    <li>cargo-workspaces</li>
    <li>git-dive</li>
    <li>git-mit</li>
    <li>nix-init</li>
    <li>projectable</li>
  </ul>
</details>

Can you merge @sternenseemann

sternenseemann commented:
> That's because cargo-audit isn't using libgit2 anymore

Checks out, libgit2-sys isn't actually being built.
